Add node list/monitoring

lohmander opened the pull request:
/nodes lists all nodes in the cluster and shows their current CPU & memory usage by polling cAdvisor. Tried make it similar to the node list in kube-ui, although a different kind of chart (as discussed in https://github.com/kubernetes/dashboard/issues/39).

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/kubernetes/dashboard/232)
<!-- Reviewable:end -->

lohmander commented:
I signed it!

maciaszczykm commented on src/app/backend/nodestats.go:
```diff
@@ -0,0 +1,164 @@
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/google/cadvisor/info/v1"
```
Could you please use `goimports` on this file? It moves external imports (`github.com/google/cadvisor/info/v1`) below these coming directly from golang (for example `time`).

bryk commented:
@lohmander Thaks for your contribution! This is a really nice piece of work. 

We need, however, to talk about this PR. Basically, we need consultation and product planning for such large features. The Dashboard needs a long term vision about what it is and what it is not and we cannot be throwing features without careful consideration. 

I believe this is not possible to make this PR into out next release (soon!), but after that - let's talk. 

@cheld

lohmander commented:
@bryk Yeah, I figured that might be the case. We needed it ourselves so that's why I just went ahead implementing it. However I suspect that other people might feel the same, so yeah, let's talk when the time is right! :)

bryk commented:
Awesome! And don't get me wrong, this is a really nice feature. We just need planning, design and discussion on new features :)

cheld commented:
@lohmander 

I checked it. Nice.

we are also about to add monitoring, but on container level:

https://github.com/kubernetes/dashboard/blob/master/docs/design/mockups/11-11-2015/16_card%20variations.png

May I ask why you added node monitoring? Did you consider using the grafana add-on? I assume it can do node monitoring out of the box

cheld commented:
@zreigz you might want to have a look at the implementation

lohmander commented:
@cheld Yeah, the container level monitoring looks really nice! 

It seemed like a nice feature to be able to quickly glance over the resource usage on the nodes from within the same web application. Also, although not relevant for most use cases, we're actually running a kubernetes cluster on a bunch of Raspberry Pis at the office so the grafana setup didn't really work, at least not out of the box. :smile:

luxas commented:
@lohmander Are you using my `kubernetes-on-arm` project? :smile:
I'm implementing `beta` dashboard support into next release

lohmander commented:
@luxas Actually no. Didn't find your repo until after we had it setup, but it sure looks nice and would've probably made the process a little bit easier. :smile:

Will be nice to have the dashboard in there from the get go!

luxas commented:
It has been on Github since 1st Oct, and I'm implementing ARM and other arch support for Kubernetes now. kubernetes/kubernetes#17981
If you like, we could discuss the setup/implementation/future for Kubernetes and ARM (via email?). 
E.g. I will probably look into the grafana addon in the future

Maybe I could help with something...

innovia commented:
@lohmander can you please fix the conflicts this i really need the monitoring and don't mind to install this version

even if you have a working image i can use that will be good too

thanks!

lohmander commented:
@innovia Hey, I've updated my branch and fixed it so that it builds again. However there seem to have been some other changes that makes it look a bit weird and unfortunately I don't have time right now to fix it. Also keep in mind that it needs to be able to access the nodes on port 4194 (cAdvisor) to get the stats if you still wanna use it. :tea:

therc commented:
Is this PR going to get revived, rewritten, or...?

bryk commented:
We can continue work on this PR now. Everything is unblocked for adding new resource types to display in the UI.

lohmander commented:
@bryk Sweet! I have been a little bit detached from the progress of this project for a while. I assume we would want to rewrite qutie a bit, like not getting node stats by polling cAdvisor on every node for instance. 

What would be a good starting point in terms of features that would need to be included?

bryk commented:
The best way to do this looks like:
1. Create an issue and describe how you'd like to show list of nodes and their details. Use ascii arts or mocks. Allow for some discussions whether the direction is good :)
2. Create a PR with backend endpoints for list of nodes. See PRs for samples - there was plenty of this recently.
3. Implement frontend that shows the list of nodes. (PR)
4. Backend for details of a node (PR)
5. Frontend for details of a node (PR).

Such scheme will make it easiest and fastest to land this feature :) Do you need any help where to start?

bryk commented:
Btw, see recent mocks for generic list/details pages: https://github.com/kubernetes/dashboard/blob/master/docs/design/navigation-and-scale.md

bryk commented:
@lohmander FYI, this is already implemented. You can access the view in `#/node` URL. I'll close the issue for now. Please comment if you need any help :)
